lib/pages/search/public_room/search_public_room_controller.dart

```diff
@@ -0,0 +1,207 @@
+import 'package:dartz/dartz.dart';
+import 'package:fluffychat/app_state/failure.dart';
+import 'package:fluffychat/app_state/success.dart';
+import 'package:fluffychat/di/global/get_it_initializer.dart';
+import 'package:fluffychat/domain/app_state/search/public_room_state.dart';
+import 'package:fluffychat/domain/usecase/search/public_room_interactor.dart';
+import 'package:fluffychat/pages/search/public_room/public_room_actions.dart';
+import 'package:fluffychat/pages/search/search_debouncer_mixin.dart';
+import 'package:fluffychat/presentation/model/search/public_room/presentation_search_public_room.dart';
+import 'package:fluffychat/presentation/model/search/public_room/presentation_search_public_room_empty.dart';
+import 'package:fluffychat/presentation/model/search/public_room/presentation_search_public_room_state.dart';
+import 'package:fluffychat/utils/dialog/twake_dialog.dart';
+import 'package:fluffychat/utils/platform_infos.dart';
+import 'package:fluffychat/utils/string_extension.dart';
+import 'package:fluffychat/utils/twake_snackbar.dart';
+import 'package:fluffychat/widgets/matrix.dart';
+import 'package:flutter/material.dart';
+import 'package:go_router/go_router.dart';
+import 'package:matrix/matrix.dart';
+import 'package:flutter_gen/gen_l10n/l10n.dart';
+
+class SearchPublicRoomController with SearchDebouncerMixin {
+  SearchPublicRoomController();
+
+  final PublicRoomInteractor _publicRoomInteractor =
+      getIt.get<PublicRoomInteractor>();
+
+  final searchResultsNotifier =
+      ValueNotifier<PresentationSearchPublicRoomUIState>(
+    PresentationSearchPublicRoomInitial(),
+  );
+
+  static const int _limitPublicRoomSearchFilter = 20;
+
+  PublicRoomQueryFilter? _filter;
+
+  bool get searchTermIsNotEmpty =>
+      _filter?.genericSearchTerm?.isNotEmpty == true;
+
+  String? get genericSearchTerm => _filter?.genericSearchTerm;
+
+  void init() {
+    initializeDebouncer((keyword) {
+      _updateFilter(keyword);
+
+      if (keyword.isEmpty) return;
+
+      if (keyword.isRoomAlias()) {
+        _resetSearchResults();
+        _searchPublicRoom();
+      } else if (keyword.isRoomId()) {
+        _handleKeywordIsRoomId();
+      }
+    });
+  }
+
+  void _updateFilter(String keyword) {
+    _filter = PublicRoomQueryFilter(
+      genericSearchTerm: keyword,
+    );
+  }
+
+  void _resetSearchResults() {
+    searchResultsNotifier.value = PresentationSearchPublicRoom(
+      searchResults: [],
+    );
+  }
+
+  void _searchPublicRoom() {
+    _publicRoomInteractor
+        .execute(
+          filter: _filter,
+          limit: _limitPublicRoomSearchFilter,
+          server: getServerName(_filter?.genericSearchTerm),
+        )
+        .listen(
+          (searchResult) => _handleListenSearchPublicRoom(searchResult),
```

Comment:
create a `streamSubscription`, then dispose it when done

Reply:
Updated